lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/ImruJobCancelledTests.cs

```diff
@@ -0,0 +1,254 @@
+using Org.Apache.REEF.Driver;
+using Org.Apache.REEF.Driver.Bridge;
+using Org.Apache.REEF.Driver.Bridge.Events;
+using Org.Apache.REEF.Driver.Evaluator;
+using Org.Apache.REEF.IMRU.API;
+using Org.Apache.REEF.IMRU.OnREEF.Driver;
+using Org.Apache.REEF.IMRU.OnREEF.Driver.StateMachine;
+using Org.Apache.REEF.IMRU.OnREEF.Parameters;
+using Org.Apache.REEF.IO.PartitionedData;
+using Org.Apache.REEF.Network.Group.Config;
+using Org.Apache.REEF.Network.Group.Driver;
+using Org.Apache.REEF.Network.Group.Driver.Impl;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Formats;
+using Org.Apache.REEF.Tang.Implementations.Configuration;
+using Org.Apache.REEF.Tang.Implementations.Tang;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Tang.Util;
+using Xunit;
+
+namespace Org.Apache.REEF.Tests.Functional.IMRU
+{
+    public class ImruJobCancelledTests
+    {
+        /// <summary>
+        /// Verifies that IMRU driver handles cancel signal: 
+        /// changes state to Fail and throw exception with "cancelled" message
+        /// </summary>
+        [Fact]
+        public void ImruDriver_HandlesCancelledEventAfterStart()
+        {
+            var driver = TangFactory
+                    .GetTang()
+                    .NewInjector(GetDriverConfig<ImruDriverTestType, ImruDriverTestType, ImruDriverTestType, ImruDriverTestType>())
+                    .GetInstance(typeof(IMRUDriver<ImruDriverTestType, ImruDriverTestType, ImruDriverTestType, ImruDriverTestType>))
+                as IMRUDriver<ImruDriverTestType, ImruDriverTestType, ImruDriverTestType, ImruDriverTestType>;
+
+            IDriverStarted startedEvent = null;
+            driver.OnNext(startedEvent);
+
+            var cancelMessage = "cancel_" + Guid.NewGuid();
+            var cancelTime = DateTime.Now;
+            IJobCancelled cancelledEvent = new JobCancelled(cancelTime, cancelMessage);
+
+            Assert.False(GetDriverState(driver).CurrentState == SystemState.Fail, "driver's state is Fail after Onstarted event");
+
+            AssertExceptionThrown<ApplicationException>(
+                () => driver.OnNext(cancelledEvent),
+                expectedExceptionMessageContent: new[] { "Job cancelled", cancelTime.ToString("u"), cancelMessage },
+                assertMessagePrefix: "Cancel event handler failed to throw expected exception");
+
+            var stateAfterCancel = GetDriverState(driver);
+            Assert.True(stateAfterCancel.CurrentState == SystemState.Fail, "invalid driver state after cancel event: expected= Fail, actual=" + stateAfterCancel.CurrentState);
+        }
+
+        private SystemStateMachine GetDriverState(object driver)
+        {
+            return driver.GetType()
+                .GetField("_systemState", BindingFlags.Instance | BindingFlags.NonPublic)
+                .GetValue(driver) as SystemStateMachine;
+        }
+
+        private void AssertExceptionThrown<TException>(Action ationWithException,
```

Comment:
This is a simple way to assert expected exception. 
            Action cancle = () => driver.OnNext(cancelledEvent);
            Assert.Throws<IMRUSystemException>(cancle);